core/schema/host.go

```diff
@@ -5,60 +5,102 @@ import (
+		dagql.Func("blob", func(ctx context.Context, parent *core.Query, args struct {
+			Digest       string `doc:"Digest of the blob"`
+			Size         int64  `doc:"Size of the blob"`
+			MediaType    string `doc:"Media type of the blob"`
+			Uncompressed string `doc:"Digest of the uncompressed blob"`
+		}) (*core.Directory, error) {
+			blobDef, err := blob.LLB(specs.Descriptor{
+				MediaType: args.MediaType,
+				Digest:    digest.Digest(args.Digest),
+				Size:      args.Size,
+				Annotations: map[string]string{
+					labels.LabelUncompressed: args.Uncompressed, // TODO ???
+				},
+			}).Marshal(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to marshal blob source: %s", err)
+			}
+			return core.NewDirectory(parent, blobDef.ToPB(), "", parent.Platform, nil), nil
+		}).Doc("Retrieves a content-addressed blob."),
+	}.Install(s.srv)
+
+	dagql.Fields[*core.Host]{
+		dagql.Func("directory", s.directory).
+			Impure().
+			Doc(`Accesses a directory on the host.`).
+			ArgDoc("path", `Location of the directory to access (e.g., ".").`).
+			ArgDoc("exclude", `Exclude artifacts that match the given pattern (e.g., ["node_modules/", ".git*"]).`).
+			ArgDoc("include", `Include only artifacts that match the given pattern (e.g., ["app/", "package.*"]).`),
+
+		dagql.Func("file", s.file).
+			Impure().
+			Doc(`Accesses a file on the host.`).
+			ArgDoc("path", `Location of the file to retrieve (e.g., "README.md").`),
+
+		dagql.Func("unixSocket", s.socket).
+			Doc(`Accesses a Unix socket on the host.`).
+			ArgDoc("path", `Location of the Unix socket (e.g., "/var/run/docker.sock").`),
+
+		dagql.Func("tunnel", s.tunnel).
+			Doc(`Creates a tunnel that forwards traffic from the host to a service.`).
+			ArgDoc("service", `Service to send traffic from the tunnel.`).
+			ArgDoc("native",
+				`Map each service port to the same port on the host, as if the service were running natively.`,
+				`Note: enabling may result in port conflicts.`).
+			ArgDoc("ports",
+				`Configure explicit port forwarding rules for the tunnel.`,
+				`If a port's frontend is unspecified or 0, a random port will be chosen
+				by the host.`,
+				`If no ports are given, all of the service's ports are forwarded. If
+				native is true, each port maps to the same port on the host. If native
+				is false, each port maps to a random port chosen by the host.`,
+				`If ports are given and native is true, the ports are additive.`),
+
+		dagql.Func("service", s.service).
+			Doc(`Creates a service that forwards traffic to a specified address via the host.`).
+			ArgDoc("ports",
+				`Ports to expose via the service, forwarding through the host network.`,
+				`If a port's frontend is unspecified or 0, it defaults to the same as
+				the backend port.`,
+				`An empty set of ports is not valid; an error will be returned.`).
+			ArgDoc("host", `Upstream host to forward traffic to.`),
+
+		dagql.Func("setSecretFile", s.setSecretFile).
```

Comment:
Yeah good catch. I think this can work the same as `host.directory()` - it can be marked `Impure()` but then do a self-call to `secret(name:...)` to return a pure Secret object so downstream things can still be cached locally.